### ISSUE TITLE
FRS Reading

### ISSUE THREAD

**Issue by cej25** [cej25/c4Root#35]
Each module currently unpacks but the data is not sorted or analysed. This is relatively urgent for FRS-coincidences and identification etc.

**Comment by cej25** [comment]
This is in progress.

Reading and Raw->Cal will be separated by Crate and/or Module, and tied together for the 2Hit ("FRS_Anl") step.

**Comment by cej25** [comment]
Reading and Raw2Cal are now mostly complete, although I believe quite inefficient for several reasons.

1) The existing Go4s (DESPEC and FRS) are quite poorly constructed, maintained, and synchronised. "Translating" them has been "fun"..

There are many, _many_ redundancies that can be optimised or outright removed IMO, although some may have to be discussed as they pertain to an element of the analysis. 

2) Attempts to optimise or correct small parts that I noticed on the go may have also been done quite poorly/incorrectly...

For example, for multi-hit detectors I would like to use vectors (as does AIDA). But I'm unclear on the difference in efficiency between:
a) creating large arrays based on some `MAX_HITS` variable, which are Zeroed every event, and output every event (filled with Zeroes when no data), and using `MAX_HITS` in every `for` loop.
b) creating vectors, which are `.clear()`d every event, output every event, and for which `.size()` is invoked for every `for` loop.

**Comment by nichubbard** [comment]
Vectors will be faster assuming MAX_HITS is large:

Array[MAX_HITS]: Each event has to write MAX_HITS 0s and loop MAX_HITS times: Consistent performance but always worst case
No memory allocation needed after creation

Vector:
clear() just sets size() to 0, single integer write 
For loop .size() means the loop is only ever for good data, for the average case where hits is 1 (probably) this means the loop is much faster
Vector never frees memory (shrinks) automatically
If not enough memory allocated, vector will dynamically increase its size, usually the capacity doubles each time it does this

You can pre-reserve a size with a vector (https://cplusplus.com/reference/vector/vector/reserve/) but after a few events it will probably stop allocating memory anyway

**Comment by nichubbard** [comment]
if i remember right ucesb the unpack to raw has some mapping.hh file where you say like

VME card 1 ch 33 is TPC41_X

maybe some mapping like that would allow flexibility? but might be a lot of work changing it all

**Comment by cej25** [comment]
Yeah this mapping functionality of UCESB was always more confusing to me, and surely to a more casual user of the code in the event they have to come and change it after the fact. I know R3B use it..

**Comment by cej25** [comment]
Pertaining to reading I think I can close this issue!
